### PR TITLE
upkeep/110: modernize developer experience

### LINKED ISSUE #110: Modernize developer experience (npm, Node, Composer, Grunt, wp/scripts)
The goal of this project is to clean up and update each 10up managed extension's npm dependencies to remove outdated packages. Additionally, we need to upgrade the Node.js version that is consistent with WooCommerce. The update process should include thorough testing to verify functionality and compatibility with the updated dependencies and Node.js version.

- [x] Upgrade to npm 10 (8 as a minimum)
- [x] Upgrade to Node 20 (including GitHub Actions)
- [x] Update Node dependencies to current major versions
- [ ] Update Composer dependencies to current major versions
- [x] Replace Grunt build processes with @wordpress/scripts
- [ ] Document npm, Node, Composer, @wordpress/scripts usage and versions in README.md

After upgrading npm/Node versions and replacing Grunt, we will ensure that e2e tests and overall plugin functionality continues to work as expected via regression testing following [defined critical flows](https://github.com/woocommerce/woocommerce-square/wiki/Critical-flows-WooCommerce-Square).

Project success metrics:

- [ ] Developer local versions are declared in the project readme file
- [ ] Extension builds properly with defined scripts and pass tests (E2E, Unit, QIT)
- [ ] PR(s) merged to trunk
- [ ] Release ZIP created and tagged for release

### PR DESCRIPTION
**Pull request by Sidsector9**
### All Submissions:

* [ ] Does your code follow the [WooCommerce Sniffs](https://github.com/woocommerce/woocommerce-sniffs/) variant of WordPress coding standards?
* [ ] Have you written new tests for your changes, as applicable?
* [ ] Have you successfully run tests with your changes locally?
* [ ] Will this change require new documentation or changes to existing documentation?

---

### Changes proposed in this Pull Request:

- Node is updated to v20.
- NPM is updated to v10.
- Node and NPM changes are reflected in the Github actions workflows.
- Build assets are generated by [wordpress/scripts](https://www.npmjs.com/package/@wordpress/scripts) only.
- Entry point for all assets are moved under `src/`

Closes #110 

### Steps to test the changes in this Pull Request:

1. Pull the branch.
2. Run `nvm use` or `nvm install` depending on the situation.
3. Verify node and NPM versions by running `node -v && npm -v`.
4. `rm -rf node_modules/`.
5. `npm ci && npm run build`.
6. Test the plugin functionality.
7. Look for any 404 errors in console.
8. Look for any errors in the debug.log file.
9. Test the final ZIP.

### Changelog entry

> Dev - Update NPM packages and node version to v20 to modernize developer experience.

### CODE REVIEW

File: src/blocks/cash-app-pay/index.js [hunk 1]
```diff
@@ -48,15 +48,15 @@ const squareCashAppPayMethod = {
 	name: PAYMENT_METHOD_ID,
 	label: <SquareCashAppPayLabel />,
 	paymentMethodId: PAYMENT_METHOD_ID,
-	ariaLabel: __('Cash App Pay payment method', 'woocommerce-square'),
-	content: <SquareComponent RenderedComponent={ComponentCashAppPay} />,
+	ariaLabel: __( 'Cash App Pay payment method', 'woocommerce' ),
```
Comment (dkotter) [inline src/blocks/cash-app-pay/index.js]: Should the textdomain be just `woocommerce` here?

File: src/new-user-experience/components/onboarding-header/index.js [hunk 2]
```diff
@@ -20,33 +20,33 @@ export const OnboardingHeader = () => {
...
-					{backStep && (
+					{ backStep && (
 						<Button
 							data-testid="previous-step-button"
-							onClick={() => setStep(backStep)}
+							onClick={ () => setStep( backStep ) }
 						>
 							<Back />
-							<span>{__('Back', 'woocommerce-square')}</span>
+							<span>{ __( 'Back', 'woocommerce' ) }</span>
```
Comment (dkotter) [inline src/new-user-experience/components/onboarding-header/index.js]:
```suggestion
							<span>{ __( 'Back', 'woocommerce-square' ) }</span>
```

File: src/new-user-experience/components/payment-gateway-settings-save-button/index.js [hunk 3]
```diff
@@ -4,11 +4,11 @@ import { check } from '@wordpress/icons';
...
 
-const withPaymentGatewaySettingsSaveButton = (WrappedComponent) => {
-	return (props) => {
+const withPaymentGatewaySettingsSaveButton = ( WrappedComponent ) => {
+	return ( props ) => {
 		const {
-			label = __('Apply Changes', 'woocommerce-square'),
-			afterSaveLabel = __('Changes Saved!'),
+			label = __( 'Apply Changes', 'woocommerce' ),
+			afterSaveLabel = __( 'Changes Saved!', 'woocommerce' ),
```
Comment (dkotter) [inline src/new-user-experience/components/payment-gateway-settings-save-button/index.js]:
```suggestion
			label = __( 'Apply Changes', 'woocommerce-square' ),
			afterSaveLabel = __( 'Changes Saved!', 'woocommerce-square' ),
```

File: src/new-user-experience/components/square-settings-save-button/index.js [hunk 4]
```diff
@@ -0,0 +1,63 @@
+import { Button } from '@wordpress/components';
+import { __ } from '@wordpress/i18n';
+import { check } from '@wordpress/icons';
+
+import { useSquareSettings } from '../../settings/hooks';
+
+const withSaveSquareSettingsButton = ( WrappedComponent ) => {
+	return ( props ) => {
+		const { label = __( 'Apply Changes', 'woocommerce' ) } = props;
```
Comment (dkotter) [inline src/new-user-experience/components/square-settings-save-button/index.js]:
```suggestion
		const { label = __( 'Apply Changes', 'woocommerce-square' ) } = props;
```

File: src/new-user-experience/components/square-settings-save-button/index.js [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+
+		const {
+			afterSaveLabel = __( 'Changes Saved!', 'woocommerce' ),
```
Comment (dkotter) [inline src/new-user-experience/components/square-settings-save-button/index.js]:
```suggestion
			afterSaveLabel = __( 'Changes Saved!', 'woocommerce-square' ),
```

File: src/new-user-experience/modules/advanced-settings/index.js [hunk 6]
```diff
@@ -30,84 +30,86 @@ export const AdvancedSettings = () => {
...
+	if ( ! squareSettingsLoaded ) {
 		return null;
 	}
 
 	return (
 		<>
 			<Section>
 				<SectionTitle
-					title={__('Advanced Settings', 'woocommerce-square')}
+					title={ __( 'Advanced Settings', 'woocommerce' ) }
```
Comment (dkotter) [inline src/new-user-experience/modules/advanced-settings/index.js]: Looks like we're changing all the textdomains here (and in other files). Curious for the reason?

File: package.json [hunk 7]
```diff
@@ -9,62 +9,37 @@
...
   },
   "devDependencies": {
-    "@babel/core": "^7.12.10",
-    "@babel/preset-env": "^7.12.11",
     "@playwright/test": "^1.44.1",
-    "@woocommerce/dependency-extraction-webpack-plugin": "^1.4.0",
+    "@woocommerce/dependency-extraction-webpack-plugin": "^3.0.1",
     "@woocommerce/eslint-plugin": "^2.2.0",
     "@wordpress/env": "^10.0.0",
     "@wordpress/scripts": "^27.9.0",
```
Comment (dkotter) [inline package.json]: Looks like the most recent version is 28.2.0 so might be nice to bump that up as part of this work

File: package.json [hunk 8]
```diff
@@ -9,62 +9,37 @@
   "title": "WooCommerce Square",
   "version": "4.7.0",
   "homepage": "https://woocommerce.com/products/woocommerce-square/",
-  "main": "Gruntfile.js",
   "scripts": {
-    "build": "composer install --no-dev && NODE_ENV=production grunt && NODE_ENV=production npm run build:webpack && npm run archive",
+    "build": "composer install --no-dev && npm run build:webpack && npm run archive",
```
Comment (dkotter) [inline package.json:13]: In testing this command, the final zip that is built contains a `assets` directory and `src` directory. Are either of these needed for the final plugin? If not, we should exclude those from the archive

**Review by dkotter — CHANGES_REQUESTED** [review]
Left a handful of comments but in testing, things seemed to work great. I see we're changing the textdomain in lots of JS files from `woocommerce-square` to just `woocommerce`. I left a handful of comments on this but then decided to quit flagging it. If this isn't intentional, we should change those back. If there is a reason why, would be curious to know.

Also, E2E tests are failing on this PR. Not sure it's related but would be great to get those fixed up so we can be confident these updates aren't causing problems.

File: src/blocks/cash-app-pay/index.js (hunk 1 above)
Comment (Sidsector9) [reply src/blocks/cash-app-pay/index.js]: I don't remember doing these changes 🤔 
I wonder if this was done by the eslint autofixer?

File: src/new-user-experience/modules/advanced-settings/index.js (hunk 6 above)
Comment (Sidsector9) [reply src/new-user-experience/modules/advanced-settings/index.js]: It's the `npm run lint:js:fix` command, I am not sure why it is replacing the text domain.

File: package.json (hunk 8 above)
Comment (Sidsector9) [reply package.json:13]: Not sure why the `assets` directory still shows after switching to this branch. But if you delete it, it doesn't show in the git diff.

File: src/new-user-experience/modules/advanced-settings/index.js (hunk 6 above)
Comment (Sidsector9) [reply src/new-user-experience/modules/advanced-settings/index.js]: For future ref: https://github.com/woocommerce/woocommerce/blob/trunk/packages/js/eslint-plugin/configs/recommended.js#L24

**Comment by Sidsector9** [comment]
@dkotter the tests are passing locally in the first pass:

Screenshot 2024-07-04 at 11 58 58 AM

I'll investigate why it is failing in pipeline.

**Review by dkotter — APPROVED** [review]

**Comment by qasumitbagthariya** [comment]
Regression / Smoke Test Report  ✅ 
-------

Tested with Archive File created via "php woorelease.phar build repo_URL" (Composer version 2.5.5, npm version 8.19.4, node version 16.20.0)

Status- Working expected with Plugin Archive/Zip file same as fix specific branch.


Next Step- Ready to Merge 🚀

**Pull request merged**